### ISSUE TITLE
Finishing a request after the connection has been closed results in an unhandled error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4013]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4013|
|Type|defect|
|Created|2009-09-10 02:49:01Z|
|Branch|https://github.com/twisted/twisted/tree/late-finish-4013|

The attached program demonstrates this.  It closes the connection the request is received over, waits a bit, then calls finish on the request object.  This results in an `AttributeError` beneath `Request.finish` as the state it relies on has been destroyed by `Request.connectionLost`.  Here's the traceback:

```
2009-09-09 22:45:24-0400 [-] Unhandled Error
        Traceback (most recent call last):
          File "requestbug.py", line 18, in <module>
            reactor.run()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/base.py", line 1166, in run
            self.mainLoop()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/base.py", line 1175, in mainLoop
            self.runUntilCurrent()
        --- <exception caught here> ---
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/base.py", line 779, in runUntilCurrent
            call.func(*call.args, **call.kw)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/web/server.py", line 239, in finish
            http.Request.finish(self)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/web/http.py", line 811, in finish
            self._cleanup()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/web/http.py", line 580, in _cleanup
            self.channel.requestDone(self)
        exceptions.AttributeError: 'NoneType' object has no attribute 'requestDone'
```

Attachments:

* [requestbug.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/abb/abb7aaaaca1e6c3a51070ebfe1e20917728f5bf6/4c0174355c8203cf2412b2e2683467854cbede46.py) (532 bytes) - added by exarkun on 2009-09-10 02:49:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4013 4013
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_late_finish_4013 branches/late-finish-4013
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1252550941000000 1252550941000000
changetime__1252952743000000 1252952743000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27320]) Branching to 'late-finish-4013'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

So I ended up just changing the exception to be more informative.  Pay attention to connection lifetimes.  I'll likely be filing another ticket shortly about some code in `twisted.web.wsgi` which does not pay sufficient attention to this.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_3"></a>|@jonathanj set owner to @exarkun|
|-|-|

Looks ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [27335]) Merge late-finish-4013

Author: exarkun
Reviewer: jonathanj
Fixes: [#4013](https://github.com/twisted/twisted/issues/4013)

Change the failure mode of `Request.finish` after `Request.connectionLost`
to raise a (tested) `RuntimeError` with a message pointing to the way to
avoid this exception, rather than bumbling along and then accidentally
raising `AttributeError`.  This should make debugging certain poorly
behaved web apps a bit easier.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
